src/verify_expr.ml

```diff
@@ -2049,32 +2126,60 @@ module VerifyExpr(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
           check_call targs h args $. fun h env retval ->
           cont (chunk::h) env retval
       end
-    | NewObject (l, cn, args) ->
+    | NewObject (l, cn, args, targsOpt) ->
       inheritance_check cn l;
       if pure then static_error l "Object creation is not allowed in a pure context" None;
-      let {cctors} = List.assoc cn classmap in
+      let {cctors; ctpenv} = List.assoc cn classmap in
       let args' = List.map (fun e -> check_expr (pn,ilist) tparams tenv e) args in
       let argtps = List.map snd args' in
-      let consmap' = List.filter (fun (sign, _) -> is_assignable_to_sign (Some pure) argtps sign) cctors in
+      let consmap' = 
+        match targsOpt with 
+          (* The new object is a parameterised type *)
+          Some(targs) -> 
+            if targs <> [] && (List.length targs) <> (List.length ctpenv) 
+            then static_error l "Invalid amount of type arguments provided" None;
+            (* No support for the diamond notation yet, so we always have the types explicitly provided *)
+            let consWithErasedSigns = List.map (fun (sign,cinfo) -> (*Iterate ovr constructors*)
+            (List.map (fun tp -> match tp with (*Iterate over arguments*)
+              RealTypeParam t -> ObjType("java.lang.Object", [])
+                | t -> t) sign, cinfo)) cctors 
+            in
+            List.filter (fun (sign, _) -> is_assignable_to_sign (Some pure) argtps sign) consWithErasedSigns
+          (* Raw type or regular object creation *)
+          | None ->  
+            let cctors = (* Possible raw type, apply type erasure to constructors signs. *)
+              List.map (fun (sign,info) -> (List.map (fun t -> match t with 
+                RealTypeParam t -> ObjType("java.lang.Object", [])
+                | t -> t) sign, info)) cctors in
+            List.filter (fun (sign, _) -> is_assignable_to_sign (Some pure) argtps sign) cctors 
+      in
       begin match consmap' with
         [] -> static_error l "No matching constructor" None
-      | [(sign, CtorInfo (lm, xmap, pre, pre_tenv, post, epost, terminates, ss, v))] ->
-        let obj = get_unique_var_symb (match xo with None -> "object" | Some x -> x) (ObjType cn) in
+      | [(sign, CtorInfo (lm, xmap, pre, pre_tenv, post, epost, terminates, ss, v, mtparams))] ->
+        (* Typecheck the type arguments *)
+        let targtps = match targsOpt with 
+          (* Type check passed type arguments *)
+          Some(targs) -> List.map (check_pure_type (pn,ilist) ctpenv Real) targs 

```

Comment:
Fout: de class type parameters zijn niet in scope in de type arguments; de type parameters van de omliggende code (= `tparams`) wel!